### PR TITLE
OboeTester: Add USB Device to Device Report

### LINKED ISSUE #1909: Add Usb Device report to OboeTester Device Report

### PR DESCRIPTION
**Pull request by robertwu1**
Fixes #1909

### CODE REVIEW

File: apps/OboeTester/app/src/main/java/com/mobileer/oboetester/DeviceReportActivity.java [hunk 1]
```diff
@@ -136,9 +141,38 @@ private void reportDeviceInfo(Collection<AudioDeviceInfo> devices) {
...
+                report.append("\nVendor ID          : 0x" + Integer.toHexString(usbDevice.getVendorId()));
+                report.append("\nDevice Name        : " + usbDevice.getDeviceName());
+                report.append("\nDevice Protocol    : " + usbDevice.getDeviceProtocol());
+                report.append("\nDevice Class       : " + usbDevice.getDeviceClass());
+                report.append("\nDevice Subclass    : " + usbDevice.getDeviceSubclass());
+                report.append("\nVersion            : " + usbDevice.getVersion());
+                report.append("\n" + usbDevice);
+                report.append("\n");
+            }
+        } catch (Exception e) {
```
[the remark this answers is not in the document]
Comment (robertwu1) [reply apps/OboeTester/app/src/main/java/com/mobileer/oboetester/DeviceReportActivity.java:168]: I don't know what exceptions could be thrown here. In general, in OboeTester, we like to catch exceptions and showing a toast. This toast makes it easier to debug if you don't have access to a debugger/laptop.

File: apps/OboeTester/app/src/main/java/com/mobileer/oboetester/DeviceReportActivity.java [hunk 2]
```diff
@@ -136,9 +141,38 @@ private void reportDeviceInfo(Collection<AudioDeviceInfo> devices) {
...
+                report.append("\nProduct Name       : " + usbDevice.getProductName());
+                report.append("\nProduct ID         : 0x" + Integer.toHexString(usbDevice.getProductId()));
+                report.append("\nManufacturer Name  : " + usbDevice.getManufacturerName());
+                report.append("\nVendor ID          : 0x" + Integer.toHexString(usbDevice.getVendorId()));
+                report.append("\nDevice Name        : " + usbDevice.getDeviceName());
+                report.append("\nDevice Protocol    : " + usbDevice.getDeviceProtocol());
+                report.append("\nDevice Class       : " + usbDevice.getDeviceClass());
+                report.append("\nDevice Subclass    : " + usbDevice.getDeviceSubclass());
+                report.append("\nVersion            : " + usbDevice.getVersion());
+                report.append("\n" + usbDevice);
```
Comment (philburk) [inline apps/OboeTester/app/src/main/java/com/mobileer/oboetester/DeviceReportActivity.java:165]: This is still dumping the massive toString() result. That makes it very hard to read the report.
Is there anything critical in the full dump?

File: apps/OboeTester/app/src/main/java/com/mobileer/oboetester/DeviceReportActivity.java [hunk 3]
```diff
@@ -136,9 +141,28 @@ private void reportDeviceInfo(Collection<AudioDeviceInfo> devices) {
...
 
+    public String reportUsbDevices() {
+        StringBuffer report = new StringBuffer();
+        report.append("\n############################");
+        report.append("\nUsb Device Report:\n");
+        try {
+            HashMap<String, UsbDevice> usbDeviceList = mUsbManager.getDeviceList();
+            for (UsbDevice usbDevice : usbDeviceList.values()) {
+                report.append("\n==== USB Device ========= " + usbDevice.getDeviceId());
+                report.append("\n" + usbDevice);
```
[the remark this answers is not in the document]
Comment (philburk) [reply apps/OboeTester/app/src/main/java/com/mobileer/oboetester/DeviceReportActivity.java:165]: Robert says the full dump contains useful information about hidden APIs like MIDI and Video.

File: apps/OboeTester/app/src/main/java/com/mobileer/oboetester/DeviceReportActivity.java (hunk 2 above)
Comment (robertwu1) [reply apps/OboeTester/app/src/main/java/com/mobileer/oboetester/DeviceReportActivity.java:165]: See UsbDevice.toString() here: https://cs.android.com/android/platform/superproject/main/+/main:frameworks/base/core/java/android/hardware/usb/UsbDevice.java;drc=cd07746e3618a2511722669b418009fc34005775;bpv=1;bpt=1;l=327?q=usbdevice

Some fields like [mHasAudioPlayback](https://cs.android.com/android/platform/superproject/main/+/main:frameworks/base/core/java/android/hardware/usb/UsbDevice.java;drc=cd07746e3618a2511722669b418009fc34005775;l=65?q=usbdevice) and [mHasMidi](https://cs.android.com/android/platform/superproject/main/+/main:frameworks/base/core/java/android/hardware/usb/UsbDevice.java;drc=cd07746e3618a2511722669b418009fc34005775;bpv=1;bpt=1;l=67?q=usbdevice&gsn=mHasMidi&gs=KYTHE%3A%2F%2Fkythe%3A%2F%2Fandroid.googlesource.com%2Fplatform%2Fsuperproject%2Fmain%2F%2Fmain%3Flang%3Djava%3Fpath%3Dandroid.hardware.usb.UsbDevice%236a70250472e668eb6e9eccb9550691d95a06d1b8b8dcee3e12c64c123b26ca1d) are only exposed through toString()

File: apps/OboeTester/app/src/main/java/com/mobileer/oboetester/DeviceReportActivity.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (flamme) [reply apps/OboeTester/app/src/main/java/com/mobileer/oboetester/DeviceReportActivity.java:168]: I understand we want to know the exceptions. But just simply catching Exception doesn't look like a good practice. And there can be exception that we don't suppress such as InterruptedException.  See [go/java-practices/interruption](https://goto.google.com/java-practices/interruption)

[the remark this answers is not in the document]
Comment (flamme) [reply apps/OboeTester/app/src/main/java/com/mobileer/oboetester/DeviceReportActivity.java:168]: Talked offline. Will keep this as is for OboeTester

**Review by flamme — APPROVED** [review]

**Pull request merged**